File: motor-dsl/runtime.py
"""
Avaliador executável + loop de runtime do motor (§22.7.7): materializa → avalia →
monitora → audita. Implementa as DUAS decisões centrais do Eixo de runtime:

  - Dois sabores de obrigação: COM prazo materializa instância (`prazo_dominio_ativo`)
    com relógio; CONTÍNUA (sem bloco prazo) não materializa — só audita.
  - `compliance_avaliacao` APPEND-ONLY = a prova de compliance (Invariante 10):
    cada avaliação carimba regra + versão do catálogo.

Ciclo da obrigação = enum FIXO em código (não template): pendente → cumprida |
vencida | dispensada | cancelada (§22.7.7, como emendas §22.4 eixo D).

Relógio é injetado (determinístico, como instante em §22.6) — sem Date.now().
Valores regulatórios (datas de prazo, feriados) são FIXTURES ilustrativos: o [GAP]
de §22.7.5 fica como GAP; aqui só exercita a forma, que independe do valor.
"""
from __future__ import annotations
from dataclasses import dataclass, field
from datetime import date, timedelta
from fractions import Fraction
from typing import Any, Dict, List, Optional, Tuple

import catalogo as cat
from nucleo import (
    Lit, ConjuntoLit, Ident, Campo, Chamada, BinOp, UnOp, parse_expr, Envelope,
)


class ErroRuntime(Exception):
    pass


# ===========================================================================
# Valores de domínio (fixtures) e tabelas-espelho do schema
# ===========================================================================
@dataclass(frozen=True)
class Competencia:
    ano: int
    mes: int

    def chave(self) -> str:
        return f"{self.ano:04d}-{self.mes:02d}"


@dataclass
class Ente:
    id: str
    populacao: int
    membros: int


@dataclass
class AtoDespesa:
    id: str
    registro_contabil: date
    publicada: bool


@dataclass
class AtoLegislativo:
    id: str
    tipo: str
    publicado: bool
    promulgacao: date


@dataclass
class Votacao:
    id: str
    materia: str
    favoraveis: int


@dataclass
class PrazoVigente:
    """Espelho de prazo_dominio_vigente (Eixo B / B4): referência regulatória."""
    jurisdicao: str
    tipo_prazo: str
    chave_periodo: str
    data_limite: date
    fonte: str
    vigente: bool


@dataclass
class Estado:
    feriados: set = field(default_factory=set)
    prazos: List[PrazoVigente] = field(default_factory=list)
    remessas: set = field(default_factory=set)   # (ente_id, sistema, competencia_key)
    bindings: Dict[str, dict] = field(default_factory=dict)  # ente_id -> {param: valor}


# ---- tabelas de runtime (espelho §22.7.7) ----
@dataclass
class ObrigacaoAtiva:
    """Espelho de prazo_dominio_ativo (polimórfico): a obrigação COM prazo."""
    id: str
    ente_id: str
    template_chave: str
    objeto_tipo: str
    objeto_id: str
    vence_em: Optional[date]
    prazo_fonte_ref: Optional[str]
    estado: str                       # enum fixo (pendente|cumprida|vencida|dispensada|cancelada)
    cumprida_em: Optional[date] = None


@dataclass(frozen=True)
class Avaliacao:
    """Espelho de compliance_avaliacao: APPEND-ONLY (sem update/delete)."""
    id: str
    ente_id: str
    obrigacao_id: Optional[str]
    template_chave: str
    registry_versao_ref: str
    veredito: str                     # conforme | nao_conforme | inaplicavel
    severidade: str
    occurred_at: date
    origem_avaliacao: str             # evento | sweep | sob_demanda
    detalhe: str = ""


# ===========================================================================
# Helpers de calendário
# ===========================================================================
def _eh_util(d: date, feriados: set) -> bool:
    return d.weekday() < 5 and d not in feriados


def _prox_dia_util(d: date, feriados: set) -> date:
    x = d + timedelta(days=1)
    while not _eh_util(x, feriados):
        x += timedelta(days=1)
    return x


def _soma_dias_uteis(d: date, n: int, feriados: set) -> date:
    x, passos = d, 0
    while passos < n:
        x += timedelta(days=1)
        if _eh_util(x, feriados):
            passos += 1
    return x


def _fim_de(comp: Competencia) -> date:
    prox = date(comp.ano + 1, 1, 1) if comp.mes == 12 else date(comp.ano, comp.mes + 1, 1)
    return prox - timedelta(days=1)


# ===========================================================================
# Avaliador de expressão (tree-walk)
# ===========================================================================
class Avaliador:
    def __init__(self, estado: Estado, agora: date):
        self.estado = estado
        self.agora = agora
        self.ultima_fonte_prazo: Optional[str] = None   # capturado p/ prazo_fonte_ref (S1/S3)

    def avaliar(self, no, amb: dict) -> Any:
        if isinstance(no, Lit):
            return no.valor
        if isinstance(no, Ident):
            if no.nome in amb:
                return amb[no.nome]
            if cat.dominio_do_literal(no.nome):
                return no.nome                     # literal de enum = seu símbolo
            raise ErroRuntime(f"identificador sem valor em runtime: {no.nome!r}")
        if isinstance(no, Campo):
            obj = self.avaliar(no.obj, amb)
            return getattr(obj, no.campo)
        if isinstance(no, ConjuntoLit):
            return frozenset(self.avaliar(e, amb) for e in no.elementos)
        if isinstance(no, Chamada):
            return self._chamada(no, amb)
        if isinstance(no, BinOp):
            return self._binop(no, amb)
        if isinstance(no, UnOp):
            v = self.avaliar(no.operando, amb)
            if no.op == "nao":
                return not v
            raise ErroRuntime(f"operador unário desconhecido: {no.op}")
        raise ErroRuntime(f"nó desconhecido: {no!r}")

    def _binop(self, no: BinOp, amb: dict) -> Any:
        op = no.op
        a = self.avaliar(no.esq, amb)
        b = self.avaliar(no.dir, amb)
        if op == "e":
            return bool(a) and bool(b)
        if op == "ou":
            return bool(a) or bool(b)
        if op == "in":
            return a in b
        if op == ">":
            return a > b
        if op == ">=":
            return a >= b
        if op == "<":
            return a < b
        if op == "<=":
            return a <= b
        if op == "==":
            return a == b
        if op == "!=":
            return a != b
        if op == "*":
            return a * b
        if op == "+":
            return a + b
        if op == "-":
            return a - b
        raise ErroRuntime(f"operador binário desconhecido: {op}")

    def _chamada(self, no: Chamada, amb: dict) -> Any:
        args = [self.avaliar(a, amb) for a in no.args]
        nome = no.nome
        fer = self.estado.feriados

        # builtins
        if nome == "hoje" or nome == "agora":
            return self.agora
        if nome == "fim_de":
            return _fim_de(args[0])
        if nome == "proximo_dia_util":
            return _prox_dia_util(args[0], fer)
        if nome == "soma_dias_uteis":
            return _soma_dias_uteis(args[0], args[1], fer)
        if nome == "arredonda_cima":
            x = args[0]
            return -((-x.numerator) // x.denominator) if isinstance(x, Fraction) else int(x)
        if nome == "fracao":
            return Fraction(args[0], args[1])      # EXATO — nunca float (armadilha T4)
        if nome == "prazo_vigente":
            return self._prazo_vigente(args[0], args[1], args[2])
        if nome == "parametro_tenant":
            ente = amb["ente"]
            binding = self.estado.bindings.get(ente.id, {})
            if args[0] not in binding:
                raise ErroRuntime(f"parametro_tenant: {args[0]!r} não configurado p/ {ente.id}")
            return binding[args[0]]

        # funções de relação
        if nome == "populacao":
            return args[0].populacao
        if nome == "membros_da_casa":
            return args[0].membros
        if nome == "remessa_enviada":
            ente, sistema, comp = args
            return (ente.id, sistema, comp.chave()) in self.estado.remessas
        if nome == "publicada_no_portal":
            return args[0].publicada
        if nome == "data_registro_contabil":
            return args[0].registro_contabil
        if nome == "publicado":
            return args[0].publicado
        if nome == "data_promulgacao":
            return args[0].promulgacao
        if nome == "votos_favoraveis":
            return args[0].favoraveis

        raise ErroRuntime(f"função sem implementação de runtime: {nome!r}")

    def _prazo_vigente(self, jurisdicao: str, tipo: str, comp: Competencia) -> date:
        chave = comp.chave()
        achados = [p for p in self.estado.prazos
                   if p.jurisdicao == jurisdicao and p.tipo_prazo == tipo
                   and p.chave_periodo == chave and p.vigente]
        if not achados:
            raise ErroRuntime(f"prazo_vigente: sem prazo p/ {jurisdicao}/{tipo}/{chave} [GAP de conteúdo]")
        self.ultima_fonte_prazo = achados[0].fonte
        return achados[0].data_limite


# ===========================================================================
# Motor: o loop materializa → avalia → monitora → audita
# ===========================================================================
LIMIAR_A_VENCER_DIAS = 5    # "a vencer" = derivação de leitura, NÃO estado persistido (§22.7.7)


class Motor:
    def __init__(self, estado: Estado, agora: date):
        self.estado = estado
        self.agora = agora
        self.av = Avaliador(estado, agora)
        self.obrigacoes: Dict[Tuple[str, str, str, str], ObrigacaoAtiva] = {}
        self.avaliacoes: List[Avaliacao] = []     # append-only
        self.eventos: List[str] = []
        self._contexto: Dict[str, Tuple[Envelope, dict]] = {}  # obrig_id -> (regra, amb) p/ resweep
        self._seq = 0

    def _id(self, prefixo: str) -> str:
        self._seq += 1
        return f"{prefixo}-{self._seq}"

    def _emitir(self, evt: str) -> None:
        self.eventos.append(evt)

    def avaliar_regra(self, regra: Envelope, registry_versao_ref: str, amb: dict,
                      objeto_tipo: str, objeto_id: str, origem: str = "evento") -> Avaliacao:
        """Um passo do loop para UMA regra contra UM objeto. Idempotente na materialização."""
        ente: Ente = amb["ente"]

        # --- avalia: aplica_quando primeiro (gatilho condicional) ---
        no_aq = parse_expr(regra.aplica_quando)
        if not self.av.avaliar(no_aq, amb):
            return self._auditar(ente.id, None, regra.template, registry_versao_ref,
                                 "inaplicavel", regra.severidade, origem,
                                 "aplica_quando=falso")

        conforme = bool(self.av.avaliar(parse_expr(regra.exige), amb))
        veredito = "conforme" if conforme else "nao_conforme"

        tem_prazo = isinstance(regra.prazo, dict) and "__malformado__" not in regra.prazo
        if not tem_prazo:
            # --- sabor CONTÍNUO: não materializa instância, só audita (§22.7.7) ---
            return self._auditar(ente.id, None, regra.template, registry_versao_ref,
                                 veredito, regra.severidade, origem,
                                 "regra contínua (sem prazo)")

        # --- sabor DEADLINE-BOUND: materializa/atualiza obrigação ---
        self.av.ultima_fonte_prazo = None
        vence_em = self.av.avaliar(parse_expr(regra.prazo["janela"]), amb)
        fonte = self.av.ultima_fonte_prazo

        chave = (ente.id, regra.template, objeto_tipo, objeto_id)
        obrig = self.obrigacoes.get(chave)
        if obrig is None:
            obrig = ObrigacaoAtiva(
                id=self._id("obr"), ente_id=ente.id, template_chave=regra.template,
                objeto_tipo=objeto_tipo, objeto_id=objeto_id,
                vence_em=vence_em, prazo_fonte_ref=fonte, estado="pendente",
            )
            self.obrigacoes[chave] = obrig
            self._contexto[obrig.id] = (regra, amb)
            self._emitir(f"ObrigacaoComplianceMaterializada({obrig.id}, {regra.template}, vence={vence_em})")
        else:
            # re-stamp do prazo só enquanto ABERTA (pendente). cumprida/vencida não movem (S3/§22.7.7).
            if obrig.estado == "pendente" and obrig.vence_em != vence_em:
                self._emitir(f"ObrigacaoComplianceReprazada({obrig.id}: {obrig.vence_em} -> {vence_em}, fonte={fonte})")
                obrig.vence_em = vence_em
                obrig.prazo_fonte_ref = fonte

        # transição de estado (enum fixo)
        if conforme:
            if obrig.estado != "cumprida":
                obrig.estado = "cumprida"
                obrig.cumprida_em = self.agora
                self._emitir(f"ObrigacaoComplianceCumprida({obrig.id})")
        else:
            if obrig.estado == "pendente" and self.agora > obrig.vence_em:
                obrig.estado = "vencida"
                self._emitir(f"ObrigacaoComplianceVencida({obrig.id})")

        return self._auditar(ente.id, obrig.id, regra.template, registry_versao_ref,
                             veredito, regra.severidade, origem, f"obrigacao={obrig.estado}")

    def _auditar(self, ente_id, obrig_id, template, registry_versao_ref,
                 veredito, severidade, origem, detalhe) -> Avaliacao:
        av = Avaliacao(
            id=self._id("aval"), ente_id=ente_id, obrigacao_id=obrig_id,
            template_chave=template, registry_versao_ref=registry_versao_ref,
            veredito=veredito, severidade=severidade or "aviso",
            occurred_at=self.agora, origem_avaliacao=origem, detalhe=detalhe,
        )
        self.avaliacoes.append(av)                # append-only
        self._emitir(f"ObrigacaoComplianceAvaliada({av.id}, {template}, {veredito})")
        return av

    # ---- monitoramento: derivação de LEITURA sobre vence_em (não persiste) ----
    def monitorar(self) -> List[dict]:
        rel = []
        for o in self.obrigacoes.values():
            if o.estado != "pendente" or o.vence_em is None:
                situacao = o.estado
            else:
                dias = (o.vence_em - self.agora).days
                if dias < 0:
                    situacao = "vencida(derivada)"
                elif dias <= LIMIAR_A_VENCER_DIAS:
                    situacao = f"a_vencer({dias}d)"
                else:
                    situacao = f"no_prazo({dias}d)"
            rel.append({"obrigacao": o.id, "template": o.template_chave,
                        "vence_em": o.vence_em, "estado": o.estado, "situacao": situacao})
        return rel

    # ---- S3: Ofício Circular desliza o prazo → re-sweep das obrigações abertas ----
    def aplicar_circular(self, jurisdicao: str, tipo_prazo: str, chave_periodo: str,
                         nova_data: date, fonte: str) -> None:
        for p in self.estado.prazos:
            if (p.jurisdicao == jurisdicao and p.tipo_prazo == tipo_prazo
                    and p.chave_periodo == chave_periodo):
                p.vigente = False
        self.estado.prazos.append(PrazoVigente(jurisdicao, tipo_prazo, chave_periodo,
                                               nova_data, fonte, vigente=True))
        self._emitir(f"PrazoDominioDeslizado({jurisdicao}/{tipo_prazo}/{chave_periodo} -> {nova_data}, {fonte})")
        # re-sweep: reavalia obrigações abertas afetadas (origem=sweep)
        for chave, obrig in list(self.obrigacoes.items()):
            if obrig.estado == "pendente" and obrig.id in self._contexto:
                regra, amb = self._contexto[obrig.id]
                self.avaliar_regra(regra, cat.CATALOGO_VERSAO, amb,
                                   obrig.objeto_tipo, obrig.objeto_id, origem="sweep")

File: prototipos/motor-dsl/src/oplenario/motor/runtime.clj
(ns oplenario.motor.runtime
  "Avaliador executável + loop de runtime do motor (§22.7.7): materializa -> avalia ->
  monitora -> audita. Implementa as DUAS decisões centrais do Eixo de runtime:

    - Dois sabores: COM prazo materializa instância (prazo_dominio_ativo) com relógio;
      CONTÍNUA (sem bloco prazo) não materializa — só audita.
    - compliance_avaliacao APPEND-ONLY = a prova de compliance (Invariante 10).

  Ciclo da obrigação = enum FIXO em código: pendente -> cumprida|vencida|dispensada|cancelada.
  Relógio injetado (determinístico, como instante em §22.6). Valores regulatórios são
  FIXTURES ilustrativos: o [GAP] de §22.7.5 segue GAP; aqui só exercita a forma.

  Estado mutável do motor vive num ATOM (read-modify-write single-thread nos testes)."
  (:require [oplenario.motor.catalogo :as cat]
            [oplenario.motor.nucleo :as nuc])
  (:import [java.time LocalDate]))

;; Cache de parse: o sweep reavalia as MESMAS strings de regra muitas vezes e parse é puro.
;; (memo não-limitado — o conjunto de strings de regra é pequeno; em produção isto vira
;;  AST pré-compilada na regra carregada. Fronteira de persistência: chat de stack.)
(def ^:private parse-memo (memoize nuc/parse-expr))

;; ===========================================================================
;; Datas + helpers de calendário
;; ===========================================================================
(defn ldate [y m d] (LocalDate/of (int y) (int m) (int d)))

(defn- util? [^LocalDate d feriados]
  (and (<= (.getValue (.getDayOfWeek d)) 5) (not (contains? feriados d))))

(defn- prox-dia-util [^LocalDate d feriados]
  (loop [x (.plusDays d 1)] (if (util? x feriados) x (recur (.plusDays x 1)))))

(defn- soma-dias-uteis [^LocalDate d n feriados]
  (loop [x d passos 0]
    (if (>= passos n) x
        (let [x2 (.plusDays x 1)] (recur x2 (if (util? x2 feriados) (inc passos) passos))))))

(defn- fim-de [comp]
  (let [{:keys [ano mes]} comp
        prox (if (= mes 12) (ldate (inc ano) 1 1) (ldate ano (inc mes) 1))]
    (.minusDays ^LocalDate prox 1)))

(defn comp-chave [comp] (format "%04d-%02d" (int (:ano comp)) (int (:mes comp))))

;; ===========================================================================
;; Construtores de fixtures (valores de domínio + tabelas-espelho do schema)
;; ===========================================================================
(defn competencia [ano mes] {:ano ano :mes mes})
(defn ente [id populacao membros] {:id id :populacao populacao :membros membros})
(defn ato-despesa [id registro-contabil publicada] {:id id :registro-contabil registro-contabil :publicada publicada})
(defn ato-legislativo [id tipo publicado promulgacao] {:id id :tipo tipo :publicado publicado :promulgacao promulgacao})
(defn votacao [id materia favoraveis] {:id id :materia materia :favoraveis favoraveis})
(defn prazo-vigente* [jurisdicao tipo-prazo chave-periodo data-limite fonte vigente]
  {:jurisdicao jurisdicao :tipo-prazo tipo-prazo :chave-periodo chave-periodo
   :data-limite data-limite :fonte fonte :vigente vigente})
(defn estado [] {:feriados #{} :prazos [] :remessas #{} :bindings {}})

;; ===========================================================================
;; Avaliador de expressão (tree-walk). ctx = {:estado :agora :fonte (atom)}.
;; ===========================================================================
(declare avaliar a-chamada a-binop arredonda-cima prazo-vigente-lookup)

(defn avaliar [no amb ctx]
  (case (:t no)
    :lit (:valor no)
    :ident (let [nm (:nome no)]
             (cond
               (contains? amb nm) (get amb nm)
               (cat/dominio-do-literal nm) nm                 ; literal de enum = seu símbolo
               :else (throw (ex-info (str "identificador sem valor em runtime: " (pr-str nm)) {:erro :runtime}))))
    :campo (get (avaliar (:obj no) amb ctx) (keyword (:campo no)))
    :conjunto-lit (set (map #(avaliar % amb ctx) (:elementos no)))
    :chamada (a-chamada no amb ctx)
    :binop (a-binop no amb ctx)
    :unop (let [v (avaliar (:operando no) amb ctx)]
            (if (= (:op no) "nao") (not v)
                (throw (ex-info (str "operador unário desconhecido: " (:op no)) {:erro :runtime}))))
    (throw (ex-info (str "nó desconhecido: " (pr-str no)) {:erro :runtime}))))

;; Duracao em runtime = mapa-tag {:duracao-dias n} (homoicônico, comparável por valor).
(defn- duracao-val? [x] (and (map? x) (contains? x :duracao-dias)))

(defn- a-mais [a b]
  (cond
    (and (number? a) (number? b)) (+ a b)
    (and (instance? LocalDate a) (duracao-val? b)) (.plusDays ^LocalDate a (long (:duracao-dias b)))
    (and (duracao-val? a) (instance? LocalDate b)) (.plusDays ^LocalDate b (long (:duracao-dias a)))
    (and (duracao-val? a) (duracao-val? b)) {:duracao-dias (+ (:duracao-dias a) (:duracao-dias b))}
    :else (throw (ex-info (str "soma inválida em runtime: " (pr-str a) " + " (pr-str b)) {:erro :runtime}))))

(defn- a-menos [a b]
  (cond
    (and (number? a) (number? b)) (- a b)
    (and (instance? LocalDate a) (duracao-val? b)) (.minusDays ^LocalDate a (long (:duracao-dias b)))
    (and (instance? LocalDate a) (instance? LocalDate b)) {:duracao-dias (- (.toEpochDay ^LocalDate a) (.toEpochDay ^LocalDate b))}
    (and (duracao-val? a) (duracao-val? b)) {:duracao-dias (- (:duracao-dias a) (:duracao-dias b))}
    :else (throw (ex-info (str "subtração inválida em runtime: " (pr-str a) " - " (pr-str b)) {:erro :runtime}))))

(defn- a-binop [no amb ctx]
  (let [op (:op no)]
    (case op
      ;; curto-circuito: o lado direito só é avaliado se necessário — regra VÁLIDA com um
      ;; ramo morto que falharia (ex.: `falso e prazo_vigente(...)`) nunca explode em runtime.
      "e"  (boolean (and (avaliar (:esq no) amb ctx) (avaliar (:dir no) amb ctx)))
      "ou" (boolean (or  (avaliar (:esq no) amb ctx) (avaliar (:dir no) amb ctx)))
      (let [a (avaliar (:esq no) amb ctx)
            b (avaliar (:dir no) amb ctx)]
        (case op
          "in" (contains? b a)
          ">"  (pos? (compare a b))
          ">=" (>= (compare a b) 0)
          "<"  (neg? (compare a b))
          "<=" (<= (compare a b) 0)
          "==" (= a b)
          "!=" (not= a b)
          "*"  (* a b)
          "+"  (a-mais a b)
          "-"  (a-menos a b)
          (throw (ex-info (str "operador binário desconhecido: " op) {:erro :runtime})))))))

(defn- arredonda-cima [x]
  (if (ratio? x)
    (quot (+ (numerator x) (dec (denominator x))) (denominator x))   ; ceil exato p/ ratio positivo
    (long x)))

(defn- prazo-vigente-lookup [ctx jurisdicao tipo comp]
  (let [chave (comp-chave comp)
        achados (filter #(and (= (:jurisdicao %) jurisdicao) (= (:tipo-prazo %) tipo)
                              (= (:chave-periodo %) chave) (:vigente %))
                        (:prazos (:estado ctx)))]
    (when (empty? achados)
      (throw (ex-info (str "prazo_vigente: sem prazo p/ " jurisdicao "/" tipo "/" chave " [GAP de conteúdo]") {:erro :runtime})))
    (reset! (:fonte ctx) (:fonte (first achados)))       ; captura p/ prazo_fonte_ref (S1/S3)
    (:data-limite (first achados))))

(defn- a-chamada [no amb ctx]
  (let [args (mapv #(avaliar % amb ctx) (:args no))
        nome (:nome no)
        st (:estado ctx)
        fer (:feriados st)]
    (case nome
      ("hoje" "agora") (:agora ctx)
      "fim_de" (fim-de (nth args 0))
      "proximo_dia_util" (prox-dia-util (nth args 0) fer)
      "soma_dias_uteis" (soma-dias-uteis (nth args 0) (nth args 1) fer)
      "arredonda_cima" (arredonda-cima (nth args 0))
      "fracao" (/ (nth args 0) (nth args 1))             ; EXATO — ratio Clojure, nunca float
      "dias" {:duracao-dias (nth args 0)}                ; construtor de Duracao
      "prazo_vigente" (prazo-vigente-lookup ctx (nth args 0) (nth args 1) (nth args 2))
      "parametro_tenant" (let [e (get amb "ente")
                               binding (get (:bindings st) (:id e) {})]
                           (if (contains? binding (nth args 0)) (get binding (nth args 0))
                               (throw (ex-info (str "parametro_tenant: " (pr-str (nth args 0)) " não configurado p/ " (:id e)) {:erro :runtime}))))
      "populacao" (:populacao (nth args 0))
      "membros_da_casa" (:membros (nth args 0))
      "remessa_enviada" (let [[e sistema comp] args]
                          (contains? (:remessas st) [(:id e) sistema (comp-chave comp)]))
      "publicada_no_portal" (:publicada (nth args 0))
      "data_registro_contabil" (:registro-contabil (nth args 0))
      "publicado" (:publicado (nth args 0))
      "data_promulgacao" (:promulgacao (nth args 0))
      "votos_favoraveis" (:favoraveis (nth args 0))
      (throw (ex-info (str "função sem implementação de runtime: " (pr-str nome)) {:erro :runtime})))))

(defn eval-expr
  "Conveniência p/ avaliar uma expressão-string solta (usado nos testes de aritmética)."
  ([s] (eval-expr s {} (estado) (ldate 2026 6 19)))
  ([s amb st agora] (avaliar (nuc/parse-expr s) amb {:estado st :agora agora :fonte (atom nil)})))

;; ===========================================================================
;; Motor: o loop materializa -> avalia -> monitora -> audita
;; ===========================================================================
(def ^:private LIMIAR-A-VENCER-DIAS 5)   ; "a vencer" = derivação de LEITURA, não estado persistido

(defn motor [estado agora]
  (atom {:estado estado :agora agora
         :obrigacoes {} :avaliacoes [] :eventos [] :contexto {} :seq 0}))

(defn- next-id! [eng prefixo]
  (str prefixo "-" (:seq (swap! eng update :seq inc))))

(defn- emitir! [eng evt] (swap! eng update :eventos conj evt))

(defn- auditar! [eng ente-id obrig-id template reg-ver veredito severidade origem detalhe]
  (let [id (next-id! eng "aval")
        av {:id id :ente-id ente-id :obrigacao-id obrig-id :template-chave template
            :registry-versao-ref reg-ver :veredito veredito :severidade (or severidade "aviso")
            :occurred-at (:agora @eng) :origem-avaliacao origem :detalhe detalhe}]
    (swap! eng update :avaliacoes conj av)             ; append-only
    (emitir! eng (str "ObrigacaoComplianceAvaliada(" id ", " template ", " veredito ")"))
    av))

(defn avaliar-regra!
  "Um passo do loop para UMA regra contra UM objeto. Idempotente na materialização."
  ([eng regra reg-ver amb objeto-tipo objeto-id]
   (avaliar-regra! eng regra reg-ver amb objeto-tipo objeto-id "evento"))
  ([eng regra reg-ver amb objeto-tipo objeto-id origem]
   (let [st (:estado @eng) agora (:agora @eng)
         e (get amb "ente") ente-id (:id e)
         ctx {:estado st :agora agora :fonte (atom nil)}]
     (if-not (avaliar (parse-memo (:aplica-quando regra)) amb ctx)
       ;; aplica_quando=falso -> inaplicável (não materializa)
       (auditar! eng ente-id nil (:template regra) reg-ver "inaplicavel" (:severidade regra) origem "aplica_quando=falso")
       (let [conforme (boolean (avaliar (parse-memo (:exige regra)) amb ctx))
             veredito (if conforme "conforme" "nao_conforme")
             prazo (:prazo regra)
             tem-prazo (and (map? prazo) (not (contains? prazo :__malformado__)))]
         (if-not tem-prazo
           ;; sabor CONTÍNUO: não materializa instância, só audita
           (auditar! eng ente-id nil (:template regra) reg-ver veredito (:severidade regra) origem "regra contínua (sem prazo)")
           ;; sabor DEADLINE-BOUND: materializa/atualiza obrigação
           (let [_ (reset! (:fonte ctx) nil)
                 vence-em (avaliar (parse-memo (get prazo "janela")) amb ctx)
                 fonte @(:fonte ctx)
                 chave [ente-id (:template regra) objeto-tipo objeto-id]
                 obrig0 (get (:obrigacoes @eng) chave)
                 obrig
                 (if (nil? obrig0)
                   (let [id (next-id! eng "obr")
                         ob {:id id :ente-id ente-id :template-chave (:template regra)
                             :objeto-tipo objeto-tipo :objeto-id objeto-id
                             :vence-em vence-em :prazo-fonte-ref fonte :estado "pendente" :cumprida-em nil}]
                     (swap! eng update :obrigacoes assoc chave ob)
                     (swap! eng update :contexto assoc id [regra amb])
                     (emitir! eng (str "ObrigacaoComplianceMaterializada(" id ", " (:template regra) ", vence=" vence-em ")"))
                     ob)
                   ;; re-stamp do prazo só enquanto ABERTA (pendente). cumprida/vencida não movem (S3).
                   (if (and (= (:estado obrig0) "pendente") (not= (:vence-em obrig0) vence-em))
                     (let [ob (assoc obrig0 :vence-em vence-em :prazo-fonte-ref fonte)]
                       (emitir! eng (str "ObrigacaoComplianceReprazada(" (:id obrig0) ": " (:vence-em obrig0) " -> " vence-em ", fonte=" fonte ")"))
                       (swap! eng update :obrigacoes assoc chave ob)
                       ob)
                     obrig0))
                 ;; transição de estado (enum fixo)
                 obrig2
                 (cond
                   conforme
                   (if (not= (:estado obrig) "cumprida")
                     (let [ob (assoc obrig :estado "cumprida" :cumprida-em agora)]
                       (emitir! eng (str "ObrigacaoComplianceCumprida(" (:id ob) ")"))
                       (swap! eng update :obrigacoes assoc chave ob) ob)
                     obrig)
                   (and (= (:estado obrig) "pendente") (.isAfter ^LocalDate agora ^LocalDate (:vence-em obrig)))
                   (let [ob (assoc obrig :estado "vencida")]
                     (emitir! eng (str "ObrigacaoComplianceVencida(" (:id ob) ")"))
                     (swap! eng update :obrigacoes assoc chave ob) ob)
                   :else obrig)]
             (auditar! eng ente-id (:id obrig2) (:template regra) reg-ver veredito (:severidade regra) origem
                       (str "obrigacao=" (:estado obrig2))))))))))

(defn monitorar
  "Derivação de LEITURA sobre vence_em (não persiste). 'a_vencer' é leitura, não estado."
  [eng]
  (let [agora (:agora @eng)]
    (mapv (fn [o]
            (let [situacao
                  (if (or (not= (:estado o) "pendente") (nil? (:vence-em o)))
                    (:estado o)
                    (let [dias (- (.toEpochDay ^LocalDate (:vence-em o)) (.toEpochDay ^LocalDate agora))]
                      (cond
                        (neg? dias) "vencida(derivada)"
                        (<= dias LIMIAR-A-VENCER-DIAS) (str "a_vencer(" dias "d)")
                        :else (str "no_prazo(" dias "d)"))))]
              {:obrigacao (:id o) :template (:template-chave o) :vence-em (:vence-em o)
               :estado (:estado o) :situacao situacao}))
          (vals (:obrigacoes @eng)))))

(defn aplicar-circular!
  "S3: Ofício Circular desliza o prazo de domínio -> re-sweep das obrigações ABERTAS."
  [eng jurisdicao tipo-prazo chave-periodo nova-data fonte]
  (swap! eng update-in [:estado :prazos]
         (fn [prazos]
           (conj (mapv (fn [p]
                         (if (and (= (:jurisdicao p) jurisdicao) (= (:tipo-prazo p) tipo-prazo)
                                  (= (:chave-periodo p) chave-periodo))
                           (assoc p :vigente false) p))
                       prazos)
                 (prazo-vigente* jurisdicao tipo-prazo chave-periodo nova-data fonte true))))
  (emitir! eng (str "PrazoDominioDeslizado(" jurisdicao "/" tipo-prazo "/" chave-periodo " -> " nova-data ", " fonte ")"))
  ;; re-sweep: reavalia obrigações abertas afetadas (origem=sweep)
  (doseq [[_chave obrig] (:obrigacoes @eng)]
    (when (and (= (:estado obrig) "pendente") (contains? (:contexto @eng) (:id obrig)))
      (let [[regra amb] (get (:contexto @eng) (:id obrig))]
        (avaliar-regra! eng regra cat/CATALOGO-VERSAO amb (:objeto-tipo obrig) (:objeto-id obrig) "sweep")))))
